src/systemctl/systemctl.c

```diff
@@ -1081,7 +1081,7 @@ static int run(int argc, char *argv[]) {
         int r;
 
         setlocale(LC_ALL, "");
-        log_parse_environment_cli();
+        log_parse_environment();
```

Comment:
> The main motivation of 41d1f46 is to not read EFI option by tools when invoked in service units, as it may be heavy or not necessary for some situations. To support them, PROC_CMDLINE_IGNORE_EFI_OPTIONS_BY_ENV flag is now set.   So, users can prohibit to read EFI options by setting SYSTEMD_PARSE_EFI=no environment variable. 

Hmm, this sounds not very user friendly. I don't think users should be involved in telling systemd tools in which context they are running. Users would suddenly have to understand what this variable does and remember to set as appropriate, which would be at least inconvenient. I think the split to have `log_parse_enviornment_cli` is appropriate: commandline utils in general *are* different then service daemons and it seems appropriate to gather settings for them in a different way.

> `systemctl reload systemd-networkd.service` drops logs in debug level even if systemd.log-level=debug is set

Shouldn't the part of the code that rereads the config be changed instead?

Reply:
> I don't think users should be involved in telling systemd tools in which context they are running.

Yeah. I agree with that. So, all executables should automatically detect in what context they are running, and we should not predefine their usecases. With the change, all executables read EFI option, kernel command line, and environment variables when run in a service. And read only environment variables when run in a terminal. This is simple, and easy to understand. no? Maybe I should split the commit into two, one is just reverting 41d1f469cf10f5f3e9cb4f4853ace9b0cfe5beae, and introduce the new flag in the other one.

Or do I misunderstand the motivation of 41d1f469cf10f5f3e9cb4f4853ace9b0cfe5beae (#16205) ? My understanding of the commit is just to optimize the startup of systemctl run in a service unit.

Reply:
> When systemctl is executed as part of some other script or service, just enabling logging for one tool (and not other tools from other projects) based on some global systemd configuration just wouldn't be useful. I expect users to be (negatively) surprised.

Conversely, I was surprised that `networkctl` does not produce debugging logs when invoked in systemd-networkd.service by `systemctl reload systemd-networkd.service` even if `systemd.log-level=debug` kernel command line is set...

Also, there exists tools like `systemd-sysusers` or `systemd-tmpfiles` which work as both daemon and cli tool... The border between daemons vs cli tools is not clear enough.

But OK. I will drop the relevant changes, at least in this PR. The main goal of this PR is dropping `LogRealm`.

>     * config should be re-read on reload

Sorry, could you elaborate more? Which config? The kernel command line options? What re-reads the config? Any commands in `ExecReload=`?

> It seems we have a bug in networkd, which we can fix without redoing how things are done in general.

`networkd`? Which issue? Or you mean, `networkctl` in `ExecReload=` I said in the above? How should we fix it without introducing 'surprise' to users? I have no idea.

Anyway, I will drop these changes from this PR, and maybe split out them to another. Thank you for the comments.

Reply:
> > When systemctl is executed as part of some other script or service, just enabling logging for one tool (and not other tools from other projects) based on some global systemd configuration just wouldn't be useful. I expect users to be (negatively) surprised.
> 
> Conversely, I was surprised that `networkctl` does not produce debugging logs when invoked in systemd-networkd.service by `systemctl reload systemd-networkd.service` even if `systemd.log-level=debug` kernel command line is set...

Yeah, that's a bit of a corner case. It would be nice to support this better.

But consider the following case: there is some service (database or whatever) that happens to call a script which does some status checks and calls `systemctl` internally. I don't think the global configuration should be read by `systemctl` when it is invoked like that.
 
> Also, there exists tools like `systemd-sysusers` or `systemd-tmpfiles` which work as both daemon and cli tool... The border between daemons vs cli tools is not clear enough.

True, this is another good case. So to cover both this and the case of `ExecReload=networkctl reload`, we should distinguish the case of commandline tools being invoked directly as service commands.

> But OK. I will drop the relevant changes, at least in this PR. The main goal of this PR is dropping `LogRealm`.

Thanks, this should help with the discussion. The other PRs should go in quickly, and then we can discuss this controversial part more easily.

> > ```
> > * config should be re-read on reload
> > ```
> 
> Sorry, could you elaborate more? Which config? The kernel command line options? What re-reads the config? Any commands in `ExecReload=`?

Any config that is read during startup. I.e. if a service reads a two different config files, looks at /proc/cmdline, and the environment, it should do the exact same things when reloading.
 
> > It seems we have a bug in networkd, which we can fix without redoing how things are done in general.
> 
> `networkd`? Which issue? Or you mean, `networkctl` in `ExecReload=` I said in the above? How should we fix it without introducing 'surprise' to users? I have no idea.

I thought you said that `networkd` looses some config on reload. That is what I meant, but I cannot find the comment now. Maybe it was a misunderstanding.

Reply:
The dropped commits are now in #18375.